### PR TITLE
Feature/params api

### PR DESCRIPTION
**Pull request by lehmannk**

### CODE REVIEW

**Comment by pathob** [comment]
Hi @lehmannk 

I added an empty plugin body to the master Branch in order to set up the CI. This is why there are some conflicts now. Don't care about them yet. We will have to create an internal Pull Request anyway in order to be able to run the Sonarqube checks then and I can help doing the rebasing and conflict handling next monday or so.

In general, there are some basic issues that have to be solved:

* Too simple commit messages: For an open source project, these commit messages are just a little bit too simple.
* Formatting: There are some formatting issues, for example curly brackets that start on a new line. This is an uncommon style for Java and will be prevented by Sonarqube soon. Please follow the automated formatting of IntelliJ ([Ctrl] + [Alt] + [L]) for now.
* Beans: Why don't you use the usual XmlElement annotations for defining the JSON serialization beans?
* Readme: Your Readme was using links for the Jira plugin. This is already solved, just use existing Readme from the master.
* Parent Pom: Please use the ASERVO parent pom as already done in the existing pom from the master.
* Atomic Transactions: Only "all or nothing" of the config should be applied to the instance if errors occur.
* CSS / JS: Can just be dropped
* Icon: We want to use the ASERVO logo

What I'm not yet sure about whether this has to be changed or not:

* Ping: The "Ping" resource is not necessary for applications with UPM. The UPM provides an endpoint to check the status of a plugin

Sounds very negative, I know. But I'm just in a hurry and of course, the first PR is the hardest.

The tests look good, but the CI will judge about it.

Kind regards
Patrick

**Comment by lehmannk** [comment]
Hi @pathob 

yeah, no problem. I just ignore the empty plugin body for now. Regarding your review comments:

> Too simple commit messages: 
totally agree, will do so on following commits. As this was more or less the initial commit I diod not find it necessary to verbose here

> Formatting:_ There are some formatting issues, for example curly brackets that start on a new line
where is that? could not find any locations in the code, can you be a bit more precises? class? line number..

> Beans: Why don't you use the usual XmlElement annotations for defining the JSON serialization beans?
to be honest, I do not see the point in using XmlElement for json processing. As the plugin using Jackson json processor all corresponding fields and classes are automatically processed. There is no need for any XmlElement annotations. Also, this makes the code cleaner and easier to read. Another suggestion is to use lombok library as I did in SettingsBean class in order to get rid of all these getter / setter methods which also helps a lot to write cleaner code

> Readme: Your Readme was using links for the Jira plugin. This is now already solved, just use now existing Readme from the master.
DONE

> Parent Pom: Please use the ASERVO parent pom as now already done in the now existing pom from the master.
DONE

> Atomic Transactions: Only "all or nothing" of the config should be applied to the instance if errors occur.
need more information here, do not understand the issue

> CSS / JS: Can just be dropped
DONE

> Icon: We want to use the ASERVO logo
DONE

> Ping: The "Ping" resource is not necessary for applications with UPM. The UPM provides an endpoint to check the status of a plugin
correct, although I have very good experience with such an easy method providing the most straight forward liveliness probe without any other system interfering. yes, we could skip this. On the other side, it's only one line of code, come on....
`public Response ping() { return Response.ok("pong").build(); }`

**Comment by pathob** [comment]
Hi @lehmannk 

> > Formatting: There are some formatting issues, for example curly brackets that start on a new line
> 
> where is that? could not find any locations in the code, can you be a bit more precises? class? line number..

I'm so confused right now. Please forget about it... :D

> > Beans: Why don't you use the usual XmlElement annotations for defining the JSON serialization beans?
> 
> to be honest, I do not see the point in using XmlElement for json processing. As the plugin using Jackson json processor all corresponding fields and classes are automatically processed. There is no need for any XmlElement annotations. Also, this makes the code cleaner and easier to read. Another suggestion is to use lombok library as I did in SettingsBean class in order to get rid of all these getter / setter methods which also helps a lot to write cleaner code

There are several points for that:

1. I want to stay as close to what Atlassian provides as possible. Using the Atlassian SDK to create the project and adding a REST module there provides code using the implicit serialization using the XmlElement annotations.
2. Three other plugins are already using these annotations.
3. It creates new code that has to be maintained (JacksonMapper, RestResource)
4. The abandonment of setters / getters / constructors just moves problem from the "beans" to the "controllers", see lines like `settings.setBaseUrl(Objects.requireNonNull(adminService.getPropertiesService().getBaseUrl()).toString());`

> > Atomic Transactions: Only "all or nothing" of the config should be applied to the instance if errors occur.
> 
> need more information here, do not understand the issue

See upcoming code comment.

> > Ping: The "Ping" resource is not necessary for applications with UPM. The UPM provides an endpoint to check the status of a plugin
> 
> correct, although I have very good experience with such an easy method providing the most straight forward liveliness probe without any other system interfering. yes, we could skip this. On the other side, it's only one line of code, come on....
> `public Response ping() { return Response.ok("pong").build(); }`

Yes, okay. :P Do you expect other util methods here? Otherwise, I would prefer just to introduce a PingResource here.

File: src/main/java/de/aservo/atlassian/bitbucket/confapi/rest/controller/AdminController.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    @PUT
+    @Consumes({MediaType.APPLICATION_JSON})
+    @Produces({MediaType.APPLICATION_JSON})
+    public Response setGeneralSettings(SettingsBean settings) {
+        try {
+            validateInputs(settings);
+            adminService.getPropertiesService().setBaseURL(URI.create(settings.getBaseUrl()));
+            adminService.getPropertiesService().setDisplayName(settings.getDisplayName());
```
Comment (pathob) [inline src/main/java/de/aservo/atlassian/bitbucket/confapi/rest/controller/AdminController.java]: "Atomic transactions": The problem is that the baseurl could be successfully set while setting the displayname could fail. The whole request would return the exception while the baseurl stays like it has been set

Comment (lehmannk) [reply src/main/java/de/aservo/atlassian/bitbucket/confapi/rest/controller/AdminController.java]: Alright, I

1. attached the @XmlElement annotations
2. removed the Jackson mapper
3. removed the utils ping controller

Regarding the RestResource class: this still performs bean validation. Implementing bean validation is more straight forward and readable IMHO than doing things like `if (name != null && name.size blahhhh)`. Thus, I suggest to keep it and maybe even integrate bean validation into the other plugins as well.

Regarding atomic operations: I am a bit confused here. Yes, in theory you are correct. However, we are talking about basic setters, not complex algorithms. Thus, risk of failure is very low plus easy to roll back for the client which needs to react on a http 500 anyway. 

Moreover look at the current Jira-Confapi-Plugin class SettingsResource, method setSettings. Isn't the same happening here? This method is not atomic either. What if setBaseUrl fails? Then still setMode will be executed.

Hopefully we can find a more pragmatic way of doing things .... ;-)

Comment (pathob) [reply src/main/java/de/aservo/atlassian/bitbucket/confapi/rest/controller/AdminController.java]: In my understanding, we don't need to care about bean / model validation. That's task of the Atlassian API. If a string is given that is too long, the Atlassian API has to return an exception and we only have to return that exception. We should not define the constraints of an interface that is not maintained by us.

Regarding the atomic operations: That's definetly something we have to take care of. There are stricter requirements to our software when we want to do open source as also topics like liability come into play. Out interfaces definitely have to meet the [ACID](https://en.wikipedia.org/wiki/ACID) principles. And you are right, the implementation of the of the Jira plugin also does not meet the ACID principles and thus has to be fixed, see https://github.com/aservo/jira-confapi-plugin/issues/6

It's not about being pragmatic or not. We all have to get used to the differences of developing open source software. The discussion for future PR's will get shorter and shorter, don't worry.

To make it short know, I will rebase your commits now and implement the last necessary changes for a first merge and open issues for the remaining tasks. We need a first merge for Sonarcloud to be running for the first time.

**Pull request closed without merging**